### PR TITLE
Truncate decimal to 4 places in alternate state visualization

### PR DESCRIPTION
**Pull request by Manvi-Agrawal**
## Code Path
`get_latex` -> `write_latex_for_term` -> `write_latex_for_real_number` -> `write_latex_for_decimal_number`

## Codespace Playground Link
https://miniature-telegram-j4vjv94qqv53pv6r-5555.app.github.dev/

## Changes Done
Truncating decimal output for state visualization for katas
image

## Scope: State Visualization in Katas
This flow doesn't impact any other outputs, like one from Python which shows 4 decimal digits accuracy.
image

### CODE REVIEW

File: compiler/qsc_eval/src/state/tests.rs
Comment (Manvi-Agrawal) [inline compiler/qsc_eval/src/state/tests.rs:571]: @tcNickolas , I am talking about these test cases. I tried out some basic superpositions, they are rendering fractions. 

Unrelated bug, for large states, it seems we omit the part of $|\psi\rangle = ... $.  

```qsharp
namespace Sample {

    open Microsoft.Quantum.Diagnostics;
    @EntryPoint()
    operation Superposition() : Unit {
        // Qubits are only accesible for the duration of the scope where they
        // are allocated and are automatically released at the end of the scope.
        
        let N=8;
        use qubit = Qubit[N];

        for i in 0..N-1{
            H(qubit[i]);
        }

        DumpMachine();

        ResetAll(qubit);
    }
}
```

File: compiler/qsc_eval/src/state/tests.rs
Comment (tcNickolas) [inline compiler/qsc_eval/src/state/tests.rs:652]: Why do you change the value used in the test itself, not its output? The number 1/4000 should also have a chance to be printed

**Comment by tcNickolas** [comment]
Looking at these test cases, I start to feel that it is indeed better to truncate the trailing zeros, showing 1 as 1.0000 does look odd. My main goal for this ask was to shorten the display of 15-digit floating-point numbers, not to lengthen the display of 1- or 2-digit ones. If the code that does that is a bit ugly, I think we can live with it - people look at the code results more often than at the code itself.

File: compiler/qsc_eval/src/state/tests.rs
Comment (billti) [inline compiler/qsc_eval/src/state/tests.rs:597]: If the value is 1, seems we'd want to show `1`, not `1.0000`.

Comment (Manvi-Agrawal) [reply compiler/qsc_eval/src/state/tests.rs]: Changed, I was kind of trying to match it with Python's output when printing state for consistency in state visualization.
image

File: compiler/qsc_eval/src/state/tests.rs
Comment (Manvi-Agrawal) [reply compiler/qsc_eval/src/state/tests.rs]: Done. Now this would be 0.0003 because of truncation to 4 decimal points. In general, I am not a big fan of testing with decimals, especially non-terminating.  I have seen weird test failures(passes on one, fails on other) due to slight change in decimal accuracy even with tolerances :-(

File: compiler/qsc_eval/src/state/tests.rs
Comment (Manvi-Agrawal) [reply compiler/qsc_eval/src/state/tests.rs]: Indeed, you were right. I kind of digged into code to find this entry point :-). Thanks @tcNickolas for clearing my doubt.
https://github.com/microsoft/qsharp/blob/28817e401664ad2ab4f53f526f37ac00e310dd69/compiler/qsc_eval/src/state.rs#L344-L348

**Review by tcNickolas — APPROVED** [review]
Looks good to me! We still need a code owner approval, though.

Thank you!

File: compiler/qsc_eval/src/state.rs
Comment (minestarks) [inline compiler/qsc_eval/src/state.rs:486]: Please refer to how the precision is set for formatting elsewhere in this file: 
https://github.com/microsoft/qsharp/blob/243ea46dd01bac3076f09a214ddb53782b09ce11/compiler/qsc_eval/src/state.rs#L27

You shouldn't have to round manually - Rust provides formatting specifiers for this purpose. `{:.4}` is the formatting specifier used to mean "4 digits after the decimal point". See Rust documentation to learn more: https://doc.rust-lang.org/std/fmt/#precision

**Review by minestarks — CHANGES_REQUESTED** [review]

**Review by minestarks — APPROVED** [review]
Please add a line comment explaining why you had to write it this way instead `{.4}` (so that it doesn't confuse other naive readers like me ;))

Solution looks good to me. Thanks for the fix!

File: compiler/qsc_eval/src/state.rs
Comment (tcNickolas) [inline compiler/qsc_eval/src/state.rs:487]:
```suggestion
        /// because we do not want trailing zeros (we need 0.5 and not 0.5000)
```

File: compiler/qsc_eval/src/state.rs
Comment (Manvi-Agrawal) [inline compiler/qsc_eval/src/state.rs:487]:
```suggestion
        // Using round() instead of neater string formatting({:.4})
        // because we do not want trailing zeros (we need 0.5 and not 0.5000)
```

**Pull request merged**
